File: frontend/src/api/jobs.ts
import { get, post } from "./index";

export interface JobInfo {
    id: string;
    name: string;
    next_run_time: string | null;
    trigger: string;
}

export interface JobsResponse {
    total: number;
    jobs: JobInfo[];
}

export interface TodayTask {
    project_name: string;
    task_type: string;
    scheduled_time: string;
    status: "pending" | "success" | "failed" | "expired";
    executed_at: string | null;
}

export interface TaskLogItem {
    id: number;
    project_name: string;
    task_type: string;
    executed_at: string | null;
    scheduled_time: string | null;
    status: string;
    error_message: string;
    task_exec_type: string;
}

export interface TaskLogsResponse {
    total: number;
    page: number;
    page_size: number;
    items: TaskLogItem[];
}

export interface ManualExecuteRequest {
    project_config_id: number;
}

export interface ManualReportDataRequest {
    project_config_id: number;
    /** 一次可入队多个 Sprint;后端按顺序排队执行(见 scheduler.py 的说明) */
    sprint_ids: string[];
}

/** 未入队的 Sprint 及原因(已在执行 / RDM 未找到) */
export interface ManualReportDataSkipped {
    sprint_id: string;
    reason: string;
}

export interface ManualReportDataResponse {
    status: string;
    /** 真正入队的 Sprint —— 轮询只盯这些,被跳过的不会产生状态 */
    task_ids: string[];
    skipped: ManualReportDataSkipped[];
    message: string;
}

export interface ReportDataStatusResponse {
    sprint_id: string;
    status: "pending" | "running" | "success" | "failed" | "unknown";
    started_at?: number;
    finished_at?: number;
    error?: string;
}

export interface TaskLogsQuery {
    page: number;
    page_size: number;
    project_name?: string;
    executed_date?: string;
    task_type?: string;
}

export const jobApi = {
    getJobs() {
        return get<JobsResponse>("/jobs")
    },

    getTodayTasks() {
        return get<TodayTask[]>("/scheduler/today-tasks")
    },

    triggerJob(jobId: string) {
        return post(`/jobs/${jobId}/trigger`)
    },

    triggerStoryReminder() {
        return post("/jobs/story-reminder/trigger")
    },

    triggerTaskReminder() {
        return post("/jobs/task-reminder/trigger")
    },

    getLogs(params: TaskLogsQuery) {
        return get<TaskLogsResponse>("/scheduler/logs", params)
    },

    manualStoryReminder(data: ManualExecuteRequest) {
        return post("/scheduler/manual/story-reminder", data)
    },

    manualTaskReminder(data: ManualExecuteRequest) {
        return post("/scheduler/manual/task-reminder", data)
    },

    manualSonarReminder(data: ManualExecuteRequest) {
        return post("/scheduler/manual/sonar-reminder", data)
    },

    /**
     * 批量检查所选 Sprint 是否已有数据。
     * 返回 `existing` = 命中的 sprint_id 列表;多选时只发一次请求(后端一次 IN 查询)。
     */
    checkReportDataExists(sprintIds: string[]) {
        return post<{ existing: string[] }>(
            "/scheduler/manual/report-data/check",
            { sprint_ids: sprintIds }
        )
    },

    manualReportData(data: ManualReportDataRequest) {
        return post<ManualReportDataResponse>(
            "/scheduler/manual/report-data",
            data
        )
    },

    getReportDataStatus(sprintId: string) {
        return get<ReportDataStatusResponse>(
            `/scheduler/manual/report-data/status/${sprintId}`
        )
    },
};

File: backend/src/api/routes/jobs.py
"""任务调度接口

收编原本直接定义在 main.py 中的 /api/jobs 系列接口,
仅依赖 get_scheduler(),不承载业务逻辑。

⚠ 整组接口都要求登录:它们会真实触发对外推送(企微/Sonar/GitLab),
  匿名可达等于把「给群里发消息」「打外部系统」开放给任何人。
"""

from fastapi import APIRouter, Depends, HTTPException

from api.auth import get_current_user_from_header
from api.scheduler import get_scheduler

router = APIRouter(
    prefix="/api/jobs",
    tags=["调度任务"],
    dependencies=[Depends(get_current_user_from_header)],
)


@router.get("")
async def get_jobs():
    """获取所有定时任务状态"""
    scheduler = get_scheduler()
    return {"total": len(scheduler.get_jobs()), "jobs": scheduler.get_jobs()}


# ⚠ 字面量路径必须注册在 /{job_id}/trigger 之前:两者段数相同,
#   否则这三个会被 {job_id} 抢先匹配,落到通用处理器里当成非法 job_id
#   (表现为 404「无效的任务类型」,接口形同不存在)。
@router.post("/story-reminder/trigger")
async def trigger_story_reminder():
    """手动触发故事提醒任务"""
    return get_scheduler().run_job_now("story_reminder")


@router.post("/task-reminder/trigger")
async def trigger_task_reminder():
    """手动触发任务提醒任务"""
    return get_scheduler().run_job_now("task_reminder")


@router.post("/sonar-reminder/trigger")
async def trigger_sonar_reminder():
    """手动触发Sonar扫描提醒任务"""
    return get_scheduler().run_job_now("sonar_reminder")


@router.post("/{job_id}/trigger")
async def trigger_job(job_id: str):
    """手动触发指定任务"""
    result = get_scheduler().run_job_now(job_id)
    if result["status"] == "error":
        raise HTTPException(status_code=404, detail=result["message"])
    return result
